lib/puppet/pops/types/string_converter.rb

```diff
@@ -835,35 +835,60 @@ def string_PStringType(val_type, val, format_map, _)
   #
   # @api public
   def puppet_quote(str)
-    if str.ascii_only? && (str =~ /(?:'|\$|\p{Cntrl}|\\)/).nil?
-      "'#{str}'"
-    else
-      bld = '"'
-      str.codepoints do |codepoint|
-        case codepoint
-        when 0x09
-          bld << '\\t'
-        when 0x0a
-          bld << '\\n'
-        when 0x0d
-          bld << '\\r'
-        when 0x22
-          bld << '\\"'
-        when 0x24
-          bld << '\\$'
-        when 0x5c
-          bld << '\\\\'
+    # Assume that the string can be single quoted
+    bld = '\''
+    escaped = false
+    str.each_codepoint do |codepoint|
+      # Control characters and non-ascii characters cannot be present in a single quoted string
+      return puppet_double_quote(str) if codepoint > 0x7f || codepoint < 0x20
+
+      if escaped
+        bld << 0x5c << codepoint
```

Comment:
Yes, and I though you complained about the one after the each. Hence my weird reply. In any case, there's no unreachable code.